### PR TITLE
[WIP] [ENH] Enhance FIR filter methods.

### PR DESCRIPTION
**Pull request by luigifcruz**
This pull-request implements the following:
- Initial conditions for the `firfilter` method.
- Zero-phase shift method called `firfilter2` (based on `scipy.signal.filtfilt`)
- FIR filter design tool `firwin2` (based on `scipy.signal.firwin2`)
- Initial conditions constructor `firfilter_zi`.
- Bindings for `lfilter` and `lfilter_zi` with `NotImplementedError` for IIR coefficients.
- Ignore `*.fatbin` files.

#### Dependencies:
This pull-request depends on the `cupy.apply_along_axis` method that is scheduled to be released in version `9.0.0`. Users can test this PR by installing `cupy==9.0.0a1` from PyPi.

#### Todo List:
- [x] Implement tests for the new methods.
- [ ] Wait for `cupy==9.0.0` release.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by awthomp** [comment]
Thanks for the PR, @luigifcruz. I love the idea and will get a bit more time to dive into your work after Thanksgiving. Expect more input around Nov 30th.

**Comment by awthomp** [comment]
ok to test

**Comment by BradReesWork** [comment]
ok to test

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by BradReesWork** [comment]
add to allowlist

**Comment by BradReesWork** [comment]
rerun tests

**Comment by awthomp** [comment]
Hey @luigifcruz -- there are some coding style changes requested by CI before tests run. You can view them [here:]( https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcusignal%2Fprb%2Fcusignal-style/detail/cusignal-style/689/pipeline)

```
>>>> FAILED: flake8 style check; begin output

./cusignal/test/test_filtering.py:547:5: F811 redefinition of unused 'TestFirfilter' from line 501

./cusignal/filtering/filtering.py:174:41: E127 continuation line over-indented for visual indent

./cusignal/filtering/filtering.py:177:53: E127 continuation line over-indented for visual indent

./cusignal/filtering/filtering.py:398:1: E305 expected 2 blank lines after class or function definition, found 1

./cusignal/filtering/filtering.py:398:1: E402 module level import not at top of file

./cusignal/filtering/filtering.py:398:1: F401 'scipy.signal as sig' imported but unused

./cusignal/filtering/filtering.py:400:1: E302 expected 2 blank lines, found 1

>>>> FAILED: flake8 style check; end output
```

**Comment by awthomp** [comment]
rerun tests

**Comment by luigifcruz** [comment]
I think the CI is installing the wrong version of Cupy. This PR requires `cupy>=9.0.0a1`. I'm not sure if I can change this. :laughing:

**Review by awthomp — COMMENTED** [review]
> I think the CI is installing the wrong version of Cupy. This PR requires `cupy>=9.0.0a1`. I'm not sure if I can change this. 😆

I'll fix this, @luigifcruz !

File: python/cusignal/filter_design/fir_filter_design.py
Comment (luigifcruz) [reply python/cusignal/filter_design/fir_filter_design.py:541]: Yes we can. I did this way because Cupy was a bit slower than Numpy. I will implement `gpumode` alongside with `cpumode`.

Comment (awthomp) [reply python/cusignal/filter_design/fir_filter_design.py:541]: I'll profile both too, and we can ultimately roll in whatever is faster. After merging in this code, we can look to optimize via custom CUDA kernels. I follow the "lets get high performance features to the public as fast as possible" and then "optimize for best perf later" montra :).

**Comment by awthomp** [comment]
Hey @luigifcruz -- as a heads up, I'm still trying to figure out how to bump the cupy version and not break CI/CD for the other RAPIDS projects. Discussion here: https://github.com/rapidsai/cusignal/pull/292

**Comment by luigifcruz** [comment]
Thanks, @awthomp! No hurry, the stable release of Cupy version 9.0.0 is meant to be released just in April.

**Comment by awthomp** [comment]
> Thanks, @awthomp! No hurry, the stable release of Cupy version 9.0.0 is meant to be released just in April. It's a shame we need to wait this long for a simple method.

Totally agree, @luigifcruz. The issue is more in how we do packaging than anything else. I'm going to sit on this PR until 9.0 is available via conda (whether a nightly release or not). Looks good, initially. I want to do some V100 and A100 testing at scale to see if we can GPU accelerate more of the internals.

I'm watching this for CuPy 9.0: https://anaconda.org/search?q=cupy

**Comment by awthomp** [comment]
rerun tests

**Comment by awthomp** [comment]
rerun tests

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
There is currently some discussion about upgrading to CuPy 9 for the other RAPIDS libraries happening on the PR below. I will temporarily request changes on this PR to prevent any accidental merges before that PR is resolved/merged.


https://github.com/rapidsai/integration/pull/205

**Comment by luigifcruz** [comment]
Great news, CuPy v9.0 was just released! https://github.com/cupy/cupy/releases/v9.0.0

**Comment by awthomp** [comment]
> Great news, CuPy v9.0 was just released! https://github.com/cupy/cupy/releases/v9.0.0

Yes! Hopefully CI/CD pulls in 9.0, and we can get this PR merged. Thanks for your patience, Luigi!

**Comment by awthomp** [comment]
rerun tests

**Comment by awthomp** [comment]
@luigifcruz We just enabled support with CuPy 9: https://github.com/rapidsai/cusignal/pull/359

I'll re-run CI today and then in the morning. I'm not totally sure how we update our conda packages (e.g. if this PR is immediate or we need to wait until a new build this evening), but I'll rerun tests again in the morning. Looking like we can get this merged soon.

**Comment by awthomp** [comment]
@ajschmidt8 -- It looks like CI tests for cusignal are still on CuPy 8.6, which is causing tests in this PR to fail. Is there anyway to force CI to pull CuPy 9.0? Looks like we're passing our builds now, though... 🤔

**Comment by ajschmidt8** [comment]
@awthomp, the containers won't have `cupy v9` until [this integration PR](https://github.com/rapidsai/integration/pull/205) and the associated PRs [here](https://github.com/rapidsai/integration/pull/205#issuecomment-827751956) are all merged. Looks like we're just waiting on https://github.com/rapidsai/cudf/pull/8082 at this point. Once that PR is merged, we can merge the `integration` PR and then all of our build/test containers will be updated to `cupy v9` and we can rerun the tests here.

**Comment by awthomp** [comment]
@luigifcruz -- Could you resubmit this PR to branch-0.20? Looks like CI will pass here in the next few days.

**Comment by ajschmidt8** [comment]
> @luigifcruz -- Could you resubmit this PR to branch-0.20? Looks like CI will pass here in the next few days.

I updated the base branch in this PR to target `branch-0.20`. should be all set now.

**Comment by awthomp** [comment]
Thanks, @ajschmidt8 !

**Comment by jakirkham** [comment]
rerun tests

(now that PR https://github.com/rapidsai/integration/pull/205 is merged :)

**Comment by awthomp** [comment]
rerun tests

**Comment by ajschmidt8** [comment]
looks like the failing checks here are leftover from the previous branch. the current checks for this branch did pass though. I'll approve this PR. @awthomp, I'll leave it to you for a final review and merge.

**Review by ajschmidt8 — APPROVED** [review]

**Comment by awthomp** [comment]
@luigifcruz -- I know it's been awhile since you've submitted this PR, and we all thank you for your patience. Before we merge, was there anything else you wanted to include here w/ CuPy 9.0?

**Comment by luigifcruz** [comment]
@awthomp -- Thanks for taking the time to review this PR! I have nothing to add at this moment.

**Review by awthomp — APPROVED** [review]

**Comment by awthomp** [comment]
@gpucibot merge

**Pull request merged**
